goalie/function_data.py

```diff
@@ -147,6 +149,135 @@ def extract(self, layout="field"):
+        if initial_condition is not None:
+            if export_field_types != ["forward"]:
+                print(
+                    "Initial condition not exported because more than 'forward' field"
+                    " type is selected for export."
+                )
+            else:
+                ics = []
+                for field, ic in initial_condition.items():
+                    ic = ic.copy(deepcopy=True)
+                    # If the function space is mixed, rename and append each
+                    # subfunction separately
+                    if hasattr(ic.function_space(), "num_sub_spaces"):
+                        for idx, sf in enumerate(ic.subfunctions):
+                            sf.rename(f"{field}_{idx}_forward")
+                            ics.append(sf)
+                    else:
+                        ic.rename(f"{field}_forward")
+                        ics.append(ic)
+                outfile.write(*ics, time=tp.subintervals[0][0])
+
+        for i in range(tp.num_subintervals):
+            for j in range(tp.num_exports_per_subinterval[i] - 1):
+                time = (
+                    tp.subintervals[i][0]
+                    + (j + 1) * tp.timesteps[i] * tp.num_timesteps_per_export[i]
+                )
+                fs = []
+                for field in tp.field_names:
+                    mixed = hasattr(self.function_spaces[field][0], "num_sub_spaces")
+                    for field_type in export_field_types:
+                        f = self._data[field][field_type][i][j].copy(deepcopy=True)
+                        if mixed:
+                            for idx, sf in enumerate(f.subfunctions):
+                                sf.rename(f"{field}_{idx}_{field_type}")
+                                fs.append(sf)
+                        else:
+                            f.rename(f"{field}_{field_type}")
+                            fs.append(f)
+                outfile.write(*fs, time=time)
+
+    def _export_h5(self, output_fpath, export_field_types, initial_condition=None):
+        """
+        Export field data to an HDF5 file. Arguments are the same as for
+        :meth:`~.export`.
+        """
+        tp = self.time_partition
+
+        # Mesh names must be unique
+        mesh_names = [fs.mesh().name for fs in self.function_spaces[tp.field_names[0]]]
+        rename_meshes = len(set(mesh_names)) != len(mesh_names)
+        with CheckpointFile(output_fpath, "w") as outfile:
+            if initial_condition is not None:
+                for field, ic in initial_condition.items():
+                    outfile.save_function(ic, name=f"{field}_initial")
+            for i in range(tp.num_subintervals):
+                if rename_meshes:
+                    mesh_name = f"mesh_{i}"
+                    msh = self.function_spaces[tp.field_names[0]][i].mesh()
+                    msh.name = mesh_name
+                    msh.topology_dm.name = mesh_name
```

Comment:
Why do you call it `msh`? `mesh` would be better.

Reply:
Bad habit :) Changed in 566a5a909a7fa9567e3c8f6ef90fb79309021787